### PR TITLE
Tktesfa2014/lnd/archv2 disag grd to topo rebase2

### PR DESCRIPTION
**Pull request by tktesfa2014**
Implementation of topography-based subgrid structure (topounits) and downscaling methods of atmospheric forcing from atmosphere grid to land topounits. It allows variable number of topounits per grid improving computational demand as compared to having the same number of topounits per grid for all grids. Downscaling methods include two approaches of precipitation downscaling developed at PNNL. Also, this branch includes downscaling methods for other atmospheric forcing developed at the University of Arizona.

### CODE REVIEW

**Comment by golaz** [comment]
@rljacob : we can include this in the v2 code base as long as it it is BFB with all current Water Cycle compsets (CMIP6 style). There will be a separate decision later whether we turn it on or not for v2.

File: components/clm/src/main/histFileMod.F90 [hunk 1]
```diff
@@ -1856,7 +1864,8 @@ subroutine htape_create (t, histrest)
 
     ! Global compressed dimensions (not including non-land points)
     call ncd_defdim(lnfid, trim(nameg), numg, dimid)
-    call ncd_defdim(lnfid, trim(namet), numt, dimid)
+    !call ncd_defdim(lnfid, trim(namet), max_topounits, dimid) !TKT topounit dimension needs to be max_topounits
```
Comment (bishtgautam) [inline components/clm/src/main/histFileMod.F90:1867]: Remove this commented out line

File: components/clm/src/main/ncdio_pio.F90.in [hunk 2]
```diff
@@ -501,11 +501,11 @@ contains
        if (ier == PIO_NOERR) nj = 1
     end if
     
-    ier = pio_inq_dimid (ncid, 'topounit', dimid)
-    if (ier == PIO_NOERR) then
-       ier = pio_inq_dimlen(ncid, dimid, ni)
-       if (ier == PIO_NOERR) nj = 1
-    end if       
+    !ier = pio_inq_dimid (ncid, 'topounit', dimid)
```
Comment (bishtgautam) [inline components/clm/src/main/ncdio_pio.F90.in:504]: Remove the commented out code

File: components/clm/src/main/subgridWeightsMod.F90 [hunk 3]
```diff
@@ -580,52 +592,86 @@ subroutine check_weights (bounds, active_only)
...
+          if (.not. weights_okay(sumwttunit(t), active_only, top_pp%active(t))) then
+             write(iulog,*) trim(subname),' ERROR: at t = ',t,'total PFT weight is ',sumwttunit(t), &
+                         'active_only = ', active_only
+             error_found = .true.
+          end if
        end if
     end do
 
     do g = bounds%begg,bounds%endg
+     !  write(iulog,*) trim(subname),' At g = ',g,'total PFT weight is ',sumwtgcell(g)
```
Comment (bishtgautam) [inline components/clm/src/main/subgridWeightsMod.F90:674]: Remove the commented out code

File: components/clm/src/main/subgridWeightsMod.F90 [hunk 4]
```diff
@@ -635,23 +681,44 @@ subroutine check_weights (bounds, active_only)
...
 
     do c = bounds%begc,bounds%endc
        l = col_pp%landunit(c)
+       t = col_pp%topounit(c)
        g = col_pp%gridcell(c)
 
        if ((active_only .and. col_pp%active(c)) .or. .not. active_only) then
           sumwtlunit(l) = sumwtlunit(l) + col_pp%wtlunit(c)
+          topo_active_only = top_pp%active(t)
+          !if(topo_active_only) then  !TKT calculate only for active topounits
```
Comment (bishtgautam) [inline components/clm/src/main/subgridWeightsMod.F90:695]: Remove the commented out code

File: components/clm/src/main/subgridWeightsMod.F90 [hunk 5]
```diff
@@ -665,23 +732,55 @@ subroutine check_weights (bounds, active_only)
 
     ! Check landunit-level weights
     sumwtgcell(bounds%begg : bounds%endg) = 0._r8
+    sumwttunit(bounds%begt : bounds%endt) = 0._r8
 
     do l = bounds%begl,bounds%endl
+       t = lun_pp%topounit(l)
        g = lun_pp%gridcell(l)
+       !topo_active_only = top_pp%active(t)
```
Comment (bishtgautam) [inline components/clm/src/main/subgridWeightsMod.F90:740]: Remove the commented out code

File: components/clm/src/main/surfrdMod.F90 [hunk 6]
```diff
@@ -596,8 +602,22 @@ subroutine surfrd_get_data (begg, endg, ldomain, lfsurdat)
...
+       call ncd_io(ncid=ncid, varname= 'TOPO2', flag='read', data=ldomain%topo2, &
+         dim1name=grlnd, readvar=readvar)
+    endif       
+        
+    !! Read the actual number of topounits per grid    
+	!call check_var(ncid=ncid, varname='topoPerGrid', vardesc=vardesc, readvar=readvar)
+    !if (readvar) then
+    !   call ncd_io(ncid=ncid, varname= 'topoPerGrid', flag='read', data=ldomain%num_tunits_per_grd, &
+    !     dim1name=grlnd, readvar=readvar)
+    !endif
```
Comment (bishtgautam) [inline components/clm/src/main/surfrdMod.F90:618]: Remove the commented out code

File: components/clm/src/main/topounit_varcon.F90 [hunk 7]
```diff
@@ -16,6 +26,104 @@ module topounit_varcon
...
   !------------------------------------------------------------------
   
-  integer, parameter, public :: max_topounits  = 1 ! maximum number of topounits per gridcell
-
+  !integer, parameter, public :: max_topounits  = 1 ! maximum number of topounits per gridcell
+  integer, public :: max_topounits               ! maximum number of topounits per gridcell
+  logical, public :: has_topounit                ! true => topounit dimension is on dataset
+  !integer, pointer, public :: ntpu_per_grd(:)             ! Number of topounits per grid globally  num_tunits_per_grd
+  !integer, pointer, public :: tpu_lnd(:)             ! Number of topounit per grid for the land domain
+  !integer, pointer, public :: tpu_glo_ind(:)         ! global index for number of topounits per grid
```
Comment (bishtgautam) [inline components/clm/src/main/topounit_varcon.F90:34]: Remove the commented out code

File: components/clm/src/main/topounit_varcon.F90 [hunk 8]
```diff
@@ -16,6 +26,104 @@ module topounit_varcon
...
+    logical :: readvar             ! read variable in or not
+    integer , allocatable :: idata2d(:,:)
+    character(len=*), parameter :: subname = 'topounit_varcon_init'
+    !-----------------------------------------------------------------------
+     
+    if (masterproc) then
+       write(iulog,*) 'Attempting to read topounit information from surface data .....'
+       if (lfsurdat == ' ') then
+          write(iulog,*)'lfsurdat must be specified'
+          !call endrun(msg=errMsg(__FILE__, __LINE__))
```
Comment (bishtgautam) [inline components/clm/src/main/topounit_varcon.F90:84]: Remove the commented out code

File: components/clm/src/main/topounit_varcon.F90 [hunk 9]
```diff
@@ -16,6 +26,104 @@ module topounit_varcon
...
+  
+       ! Make sure the number of topounits per grid consistent with the land mask
+       do n = begg,endg
+          !write(iulog,*)'TKT num_tunits_per_grd(n), ldomain%mask(n) ', ldomain%num_tunits_per_grd(n), '<====>', ldomain%mask(n)
+          if (ldomain%num_tunits_per_grd(n)>1 .and. ldomain%mask(n)==0) then
+             write(iulog,*)trim(subname),&
+                  'initialize1: landmask/Number of topounits mismatch'
+             write(iulog,*)trim(subname),&
+                  'More than 1 topounits where landmask = 0, gridcell index', n
+             !call endrun(msg=errMsg(__FILE__, __LINE__))
```
Comment (bishtgautam) [inline components/clm/src/main/topounit_varcon.F90:122]: Remove the commented out code

File: components/clm/src/biogeophys/CanopyFluxesMod.F90 [hunk 10]
```diff
@@ -1125,6 +1126,12 @@ subroutine CanopyFluxes(bounds,  num_nolakeurbanp, filter_nolakeurbanp, &
            p = filterp(f)
            t = veg_pp%topounit(p)
            lbl_rsc_h2o(p) = getlblcef(forc_rho(t),t_veg(p))*uaf(p)/(uaf(p)**2._r8+1.e-10_r8)   !laminar boundary resistance for h2o over leaf, should I make this consistent for latent heat calculation?
+           !if (masterproc) then  ! TKT debugging
+           !   write (iulog,*) 'forc_rho(t): ', forc_rho(t)
+           !   write (iulog,*) 't_veg(p): ', t_veg(p)
+           !   write (iulog,*) 'uaf(p): ', uaf(p)
+           !   write (iulog,*) 'uaf(p)**2._r8+1.e-10_r8: ', uaf(p)**2._r8+1.e-10_r8   
+           !end if
```
Comment (bishtgautam) [inline components/clm/src/biogeophys/CanopyFluxesMod.F90:1134]: Remove the commented out code

File: components/clm/src/cpl/lnd_disagg_forc.F90 [hunk 11]
```diff
@@ -0,0 +1,690 @@
...
+       if (precip_downscaling_method == 'FNM') then
+          allocate(deltaRain(numt_pg))
+          deltaRain(:) = 0._r8
+          allocate(deltaSnow(numt_pg))
+          deltaSnow(:) = 0._r8
+          hrise = 0.
+       end if
+    
+       do t = grc_pp%topi(g), grc_pp%topf(g)
+       !do t = 1, numt_pg                                 !loop through the valid topounits only
```
Comment (bishtgautam) [inline components/clm/src/cpl/lnd_disagg_forc.F90:159]: Remove the commented out code

File: components/clm/src/cpl/lnd_disagg_forc.F90 [hunk 12]
```diff
@@ -0,0 +1,690 @@
...
+                top_af%snow(t) = snow_g
+             else
+                top_af%rain(t) = rain_g + (deltaRain(t2) - (rain_g/mxElv)*(sum_of_hrise/numt_pg))
+                top_af%snow(t) = snow_g + (deltaSnow(t2) - (snow_g/mxElv)*(sum_of_hrise/numt_pg))
+             end if
+             !if (masterproc) then  ! TKT debugging
+             !   write(iulog,*) ' top_af%rain(t) = ', top_af%rain(t)
+             !   write(iulog,*) ' top_af%snow(t) = ', top_af%snow(t)
+             !   write(iulog,*) ' mxElv =  ', mxElv
+             !end if
```
Comment (bishtgautam) [inline components/clm/src/cpl/lnd_disagg_forc.F90:228]: Remove the commented out code

File: components/clm/src/cpl/lnd_disagg_forc.F90 [hunk 13]
```diff
@@ -0,0 +1,690 @@
...
+       lwrad_t = lwrad_g + lnd2atm_vars%eflx_lwrad_out_grc(g) * &
+          4._r8 * (tair_t - tair_g) / tsfc_g
+!       if (masterproc) then  ! TKT debugging
+!            write(iulog,*) ' lwrad_g =  ', lwrad_g 
+!            write(iulog,*) ' lnd2atm_vars%eflx_lwrad_out_grc(g) ', lnd2atm_vars%eflx_lwrad_out_grc(g) 
+!            write(iulog,*) ' tair_t ',tair_t
+!            write(iulog,*) ' tair_g ',tair_g
+!            write(iulog,*) ' tsfc_g ',tsfc_g
+!            write(iulog,*) ' lwrad_t ',lwrad_t
+!       end if
```
Comment (bishtgautam) [inline components/clm/src/cpl/lnd_disagg_forc.F90:496]: Remove the commented out code

File: components/clm/src/cpl/lnd_disagg_forc.F90 [hunk 14]
```diff
@@ -0,0 +1,690 @@
...
+
+       ! Make sure that, after normalization, the grid cell mean is conserved
+!           if (sum_wts_g > 0._r8) then
+!               if (abs((newsum_lwrad_g / sum_wts_g) - lwrad_g) > 1.e-8_r8) then
+!                  write(iulog,*) 'g, newsum_lwrad_g, sum_wts_g, forc_lwrad_g: ', &
+!                       g, newsum_lwrad_g, sum_wts_g, lwrad_g
+!                  call endrun(msg=' ERROR: Energy conservation error downscaling longwave'//&
+!                       errMsg(__FILE__, __LINE__))
+!               end if
+!            end if
```
Comment (bishtgautam) [inline components/clm/src/cpl/lnd_disagg_forc.F90:512]: Remove the commented out code

File: components/clm/src/cpl/lnd_disagg_forc.F90 [hunk 15]
```diff
@@ -0,0 +1,690 @@
...
+    !   write(iulog,*) ' uovern_t =  ', uovern_t 
+    !   write(iulog,*) ' grdElv ', grdElv 
+    !   write(iulog,*) ' topoElv ', topoElv       
+    !   write(iulog,*) ' mxElv ', mxElv
+    !   write(iulog,*) ' rain_g ', rain_g
+    !   write(iulog,*) ' snow_g ', snow_g
+    !   write(iulog,*) ' hrise ', hrise
+    !   write(iulog,*) ' deltaR ', deltaR
+    !   write(iulog,*) ' deltaS ', deltaS
+    !end if
```
Comment (bishtgautam) [inline components/clm/src/cpl/lnd_disagg_forc.F90:577]: Remove the commented out code

File: components/clm/src/main/decompInitMod.F90 [hunk 16]
```diff
@@ -442,38 +456,40 @@ subroutine decompInit_clumps(glcmask)
...
+          procinfo%ntunits  = procinfo%ntunits  + itunits
           procinfo%nlunits     = procinfo%nlunits  + ilunits
           procinfo%ncols       = procinfo%ncols    + icols
           procinfo%npfts       = procinfo%npfts    + ipfts
           procinfo%nCohorts    = procinfo%nCohorts + icohorts
        endif
 
     enddo
-
+    !if (masterproc) then
```
Comment (bishtgautam) [inline components/clm/src/main/decompInitMod.F90:490]: Please remove commented out lines

File: components/clm/src/main/initGridCellsMod.F90 [hunk 17]
```diff
@@ -123,78 +126,110 @@ subroutine initGridcells
...
+                     is_tpu_active = .true.
+                 else
+                     is_tpu_active = .false.
+                 endif                    
+             endif
+             elv = elv_tunit(gdc,topounit) !grc_pp%televation(gdc,topounit) 
+             slp = slp_tunit(gdc,topounit) !grc_pp%tslope(gdc,topounit) 
+             asp = asp_tunit(gdc,topounit) !grc_pp%taspect(gdc,topounit) 
+             topo_ind = topounit
+             !wttopounit2gridcell = 1._r8/(max_topounits)
```
Comment (bishtgautam) [inline components/clm/src/main/initGridCellsMod.F90:157]: Please remove commented out lines

**Review by bishtgautam — CHANGES_REQUESTED** [review]
I only reviewed changes in the land model and request minor changes.

File: components/clm/src/main/controlMod.F90 [hunk 18]
```diff
@@ -879,6 +882,9 @@ subroutine control_print ()
     write(iulog,*) 'input data files:'
     write(iulog,*) '   PFT physiology and parameters file = ',trim(paramfile)
     write(iulog,*) '   Soil order dependent parameters file = ',trim(fsoilordercon)
+    write(iulog,*) '    use_downscaling_to_topounit = ', use_downscaling_to_topounit
+    write(iulog,*) '    precip_downscaling_method = ', precip_downscaling_method
+    !write(iulog,*) '    Maximum number of topounits per grid = ', max_topounits
```
[the remark this answers is not in the document]
Comment (tktesfa2014) [reply components/clm/src/main/controlMod.F90]: Thanks @bishtgautam . Comment removed.

File: components/clm/src/main/topounit_varcon.F90 [hunk 19]
```diff
@@ -16,6 +26,104 @@ module topounit_varcon
...
+    if (.not. has_topounit) then
+       max_topounits = 1
+       if (masterproc) then          
+          write(iulog,*)'Surface dataset has no topounit dimention; max_topounits is set to 1'
+       end if
+    else
+       call ncd_inqdlen(ncid, dimid, topounits_size)  ! Get the dimension size of topounit from file
+       max_topounits = topounits_size
+    endif
+    if(has_topounit .and. max_topounits > 1) then
```
[the remark this answers is not in the document]
Comment (tktesfa2014) [reply components/clm/src/main/topounit_varcon.F90]: Thanks! Code has been changed as recommended.

File: components/clm/src/main/histFileMod.F90 (hunk 1 above)
Comment (tktesfa2014) [reply components/clm/src/main/histFileMod.F90]: Thanks! Commented line has been removed.

File: components/clm/src/main/ncdio_pio.F90.in (hunk 2 above)
Comment (tktesfa2014) [reply components/clm/src/main/ncdio_pio.F90.in]: Thanks! Commented code has been removed.

File: components/clm/src/main/subgridWeightsMod.F90 (hunk 3 above)
Comment (tktesfa2014) [reply components/clm/src/main/subgridWeightsMod.F90]: Thanks! Commented out code has been removed.

File: components/clm/src/main/decompInitMod.F90 (hunk 16 above)
Comment (tktesfa2014) [reply components/clm/src/main/decompInitMod.F90]: Thanks! Commened out code has been removed.

File: components/clm/src/main/initGridCellsMod.F90 (hunk 17 above)
Comment (tktesfa2014) [reply components/clm/src/main/initGridCellsMod.F90]: @bishtgautam thanks! Commented out code has been removed.

Comment (tktesfa2014) [reply components/clm/src/main/initGridCellsMod.F90]: Thanks @bishtgautam ! I have addressed all your comments in the latest commit.

**Comment by tktesfa2014** [comment]
I have addressed all comments from @bishtgautam . Thanks!

File: components/clm/src/main/topounit_varcon.F90 (hunk 19 above)
[the remark this answers is not in the document]
Comment (bishtgautam) [reply components/clm/src/main/topounit_varcon.F90]: There are a few more locations in the code where `if(has_topounit .and. max_topounits > 1)` is used. Could you please make a similar change?

File: components/clm/src/cpl/lnd_disagg_forc.F90 (hunk 15 above)
Comment (tktesfa2014) [reply components/clm/src/cpl/lnd_disagg_forc.F90]: Thanks! Commented out code has been removed.

File: components/clm/src/cpl/lnd_disagg_forc.F90 (hunk 14 above)
Comment (tktesfa2014) [reply components/clm/src/cpl/lnd_disagg_forc.F90]: Thanks! Commented out code has been removed.

File: components/clm/src/cpl/lnd_disagg_forc.F90 (hunk 13 above)
Comment (tktesfa2014) [reply components/clm/src/cpl/lnd_disagg_forc.F90]: Thanks! Commented out code has been removed.

File: components/cam/src/physics/cam/froude.F90 [hunk 20]
```diff
@@ -0,0 +1,176 @@
...
+          end do
+          do k = 1, pver
+            pdel(k) = pint(k+1) - pint(k)
+            pmid(k) = 0.5_r8*(pint(k) + pint(k+1))
+            t(k)=state%t(i,k)
+            q(k)=state%q(i,k,1)
+          end do
+          kmax=pver-1
+          do k=kmax-1,1,-1
+             if(pint(kmax)-pint(k).gt.2.e4)then
```
Comment (singhbalwinder) [inline components/cam/src/physics/cam/froude.F90:88]: I think we need to change 2.e4 to 2.e4_r8 (or 2.0e4_r8)

File: components/cam/src/physics/cam/froude.F90 [hunk 21]
```diff
@@ -0,0 +1,176 @@
...
+            call qsat(t(k),pmid(k),es,qs)
+            rh(k)=q(k)/qs
+          end do
+!
+!---- determine the brunt-vaisala frequency
+!
+          do k = 2, pver
+            thx(k)=0.5*(state%t(i,k)*(state%ps(i)/pmid(k))**cappa+state%t(i,k-1)*(state%ps(i)/pmid(k-1))**cappa)
+            tint(k)=0.5*(state%t(i,k)+state%t(i,k-1))
+            qint(k)=0.5*(state%q(i,k,1)+state%q(i,k-1,1))
```
Comment (singhbalwinder) [inline components/cam/src/physics/cam/froude.F90:104]: Same for these 0.5. Change them to 0.5_r8.

File: components/cam/src/physics/cam/froude.F90 [hunk 22]
```diff
@@ -0,0 +1,176 @@
...
+          tint(pver+1) = state%t(i,pver)
+          qint(pver+1) = state%q(i,pver,1)
+!
+          zs = state%phis(i) / gravit
+          zint(pver+1) = zs
+!
+          rovg=rair/gravit
+          do k = pver, 1, -1
+            zint(k) = zint(k+1)+                             &
+                 rovg*state%t(i,k)*(1. + zvir * state%q(i,k,1))*(lnpint(k+1)-lnpint(k))
```
Comment (singhbalwinder) [inline components/cam/src/physics/cam/froude.F90:117]: Same for 1. . It should be 1.0_r8

File: components/cam/src/physics/cam/froude.F90 [hunk 23]
```diff
@@ -0,0 +1,176 @@
...
+            zmid(k) = 0.5_r8*(zint(k)+zint(k+1))
+          end do
+        !  zmid(pver+1) = zint(pver+1) 
+          zmid(pver) = zint(pver+1) !TKT
+!
+!---- maximum height rise of an air parcel
+!
+          bvfmin = 1.e5_r8
+          do k = kmin,kmax
+            if(rh(k).lt.0.9)then
```
Comment (singhbalwinder) [inline components/cam/src/physics/cam/froude.F90:130]: same here: 0.9_r8

File: components/cam/src/physics/cam/froude.F90 [hunk 24]
```diff
@@ -0,0 +1,176 @@
...
+            bvfmin=min(bvf(k),bvfmin)
+
+          end do
+          if(bvfmin <= 0.0_r8) then
+            uovern(i) = 1.e8_r8
+          else
+            uovern(i) = 0.0_r8
+            cmass = 0.0_r8
+            wspdbar=0
+            bvfbar=0
```
Comment (singhbalwinder) [inline components/cam/src/physics/cam/froude.F90:151]: Both wspbar and bvfbar should be initialized to 0.0_r8

File: components/eam/src/physics/cam/froude.F90 [hunk 25]
```diff
@@ -0,0 +1,176 @@
+MODULE froude
```
Comment (singhbalwinder) [inline components/eam/src/physics/cam/froude.F90:1]: This is the file which should be kept. I didn't realize the other file I commented on was in "components/cam" directory, which should be deleted. Please follow the comments I made to that file to fix issues in this file.

**Review by singhbalwinder — CHANGES_REQUESTED** [review]
Thanks Teklu! This seems like a lot of work. I have some minor comments.

There are a few things I noticed in this PR:
1. Submodules has been accidentally removed from both ELM and EAM, which should be restored.
2. Old "cam" and "clm" directories should be removed

File: components/clm/src/main/topounit_varcon.F90 (hunk 19 above)
[the remark this answers is not in the document]
Comment (tktesfa2014) [reply components/clm/src/main/topounit_varcon.F90]: Thanks @bishtgautam . All requested changes have been applied in the recent commit.

File: components/cam/src/physics/cam/froude.F90 (hunk 20 above)
Comment (tktesfa2014) [reply components/cam/src/physics/cam/froude.F90:88]: Thanks @singhbalwinder ! All requested changes have been applied in the latest commit.

File: components/cam/src/physics/cam/froude.F90 (hunk 21 above)
Comment (tktesfa2014) [reply components/cam/src/physics/cam/froude.F90:104]: Thanks! Changes have been made.

File: components/cam/src/physics/cam/froude.F90 (hunk 22 above)
Comment (tktesfa2014) [reply components/cam/src/physics/cam/froude.F90:117]: Thanks! Requested changes have been applied.

File: components/cam/src/physics/cam/froude.F90 (hunk 23 above)
Comment (tktesfa2014) [reply components/cam/src/physics/cam/froude.F90:130]: Thanks! Changes have been applied as requested.

File: components/cam/src/physics/cam/froude.F90 (hunk 24 above)
Comment (tktesfa2014) [reply components/cam/src/physics/cam/froude.F90:151]: Thanks! Requested changes have been applied.

File: components/eam/src/physics/cam/froude.F90 (hunk 25 above)
Comment (tktesfa2014) [reply components/eam/src/physics/cam/froude.F90:1]: Thanks! Requested changes have been applied.

**Comment by tktesfa2014** [comment]
Thanks @bishtgautam  and @singhbalwinder , All requested changes have been addressed in the latest commit.

File: components/elm/src/dyn_subgrid/do_genf90 [hunk 26]
```diff
@@ -2,7 +2,8 @@
 
 # Run genf90 on all files that need it in this directory
 
-genf90=../../../../../../tools/cprnc/genf90/genf90.pl
+# genf90=../../../../../../tools/cprnc/genf90/genf90.pl
```
Comment (thorntonpe) [inline components/elm/src/dyn_subgrid/do_genf90:5]: Remove commented code?

**Review by thorntonpe — APPROVED** [review]
Teklu, I reviewed all changes and had only 1 suggestion to remove some commented code. You've undertaken a major development effort, and the resulting code is clear and logically organized. This brings an important new capability, and I'm looking forward to using it for new science.

**Comment by rljacob** [comment]
@tktesfa2014 please fix the title of the PR to describe what it does.  Is this BFB?  Needed for v2?

**Comment by golaz** [comment]
@rljacob, realistically, this is unlikely to be used for v2 Water Cycle, but maybe for BGC. If it's BFB, I don't have any problem having it merged on master.

**Review by singhbalwinder — APPROVED** [review]
Everything looks good to me. I think we need to rebase (or master merge) it before merging it to master as there are some files which are in conflict.

**Comment by rljacob** [comment]
Notes from IG telecon:  will be rebased.

**Comment by rljacob** [comment]
Waiting for @tktesfa2014 to rebase.

**Pull request closed without merging**